### PR TITLE
Added workset calculator utility class, extended statistics by workset

### PR DESCRIPTION
**Pull request by tomaszrybicki**
Signed-off-by: Tomasz Rybicki <tomasz.rybicki@intel.com>

### CODE REVIEW

**Comment by tomaszrybicki** [comment]
Tests to be added

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * Copyright(c) 2012-2018 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+#include <octf/analytics/statistics/WorksetCalculator.h>
+
+namespace octf {
+
+WorksetCalculator::Range::Range() {
+    this->begin = 0;
+    this->end = 0;
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: Use initializer lists

File: tests/octf/analytics/statistics/WorksetCalculatorTest.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    int ranges = 100000;
+    int range = 0;
+
+    int len = 10;
+    int begin = 0;
+
+    // Add adjacent ranges like so:
+    // [0-10], [10-20], [20-30], ...
+    for (; range < ranges; begin += (len - 1), range++) {
```
Comment (tomaszrybicki) [inline tests/octf/analytics/statistics/WorksetCalculatorTest.cpp]: just begin += len

File: tests/octf/analytics/statistics/WorksetCalculatorTest.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        wc.insertRange(begin, len);
+    }
+
+    ASSERT_EQ(wc.getWorkset(), ranges * len - ranges * 1 + 1);
+}
+
+TEST(WorksetCalculator, overlappingRanges) {
```
Comment (tomaszrybicki) [inline tests/octf/analytics/statistics/WorksetCalculatorTest.cpp]: just ranges * len

File: tests/octf/analytics/statistics/WorksetCalculatorTest.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    int len = 10;
+    int begin = 0;
+
+    // Add adjacent ranges like so:
+    // [0-10], [10-20], [20-30], ...
+    for (; range < ranges; begin += (len - 1), range++) {
+        wc.insertRange(begin, len);
+    }
+
+    ASSERT_EQ(wc.getWorkset(), ranges * len - ranges * 1 + 1);
```
Comment (tomaszrybicki) [inline tests/octf/analytics/statistics/WorksetCalculatorTest.cpp]: just ranges*len

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+WorksetCalculator::Range::Range(const Range &other) {
+    this->begin = other.begin;
+    this->end = other.end;
+}
+
+WorksetCalculator::Range &WorksetCalculator::Range::operator=(
+        const Range &other) {
+    this->begin = other.begin;
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: Good practice is check if this is different then other then perform assignment

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    this->begin = 0;
+    this->end = 0;
+}
+
+WorksetCalculator::Range::Range(uint64_t _begin, uint64_t _end) {
+    this->begin = _begin;
+    this->end = _end;
+}
+
+WorksetCalculator::Range::Range(const Range &other) {
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: Use initialization list

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    if (newRange.begin >= overlapIter->begin &&
+        newRange.end <= overlapIter->end) {
+        return;
+    }
+
+    // Delete the overlapping ranges
+    overlapIter = m_hitRanges.erase(overlapIter);
+
+    while (overlapIter != m_hitRanges.end() &&
+           Range::doRangesOverlap(newRange, *overlapIter)) {
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: You can equal operator

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        if (newRange.begin >= overlapIter->begin &&
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: dead code, can be removed

File: tests/octf/analytics/statistics/WorksetCalculatorTest.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    // Insert the biggest range
+    wc.insertRange(currentBegin, maxLen);
+
+    // Add subranges like so:
+    // First range: [0-100]
+    // Next ranges: [1-99], [2-98], ...
+    for (; range < ranges; range++) {
+        currentBegin++;
+        currentLen--;
+        assert(currentLen > 0);
```
Comment (mariuszbarczak) [inline tests/octf/analytics/statistics/WorksetCalculatorTest.cpp]: Maybe we should use gtest assert

File: tests/octf/analytics/statistics/WorksetCalculatorTest.cpp [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+    int ranges = 100000;
+    int range = 0;
+
+    uint64_t len;
+    uint64_t begin;
+
+    for (; range < ranges; range++) {
+        begin = rand(rng) - 1;
+        len = rand(rng);
```
Comment (mariuszbarczak) [inline tests/octf/analytics/statistics/WorksetCalculatorTest.cpp]: I'd add two test, first with big len, second with small

File: source/octf/analytics/statistics/WorksetCalculator.cpp (hunk 7 above)
Comment (tomaszrybicki) [reply source/octf/analytics/statistics/WorksetCalculator.cpp]: I think it is more clear this way

File: tests/octf/analytics/statistics/WorksetCalculatorTest.cpp (hunk 9 above)
Comment (tomaszrybicki) [reply tests/octf/analytics/statistics/WorksetCalculatorTest.cpp]: I think it is an assert on the test code and not on the tested code and should be left as it is.

File: source/octf/analytics/statistics/IoStatistics.cpp [hunk 11]
```diff
@@ -12,7 +12,8 @@ struct IoStatistics::Stats {
     Stats()
             : SizeDistribution("sector", 4096, 2)
             , LatencyDistribution("ns", 100, 10)
-            , Errors(0) {}
+            , Errors(0)
+            , Wc() {}
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/IoStatistics.cpp]: Please consider Wc member in other constructors

File: source/octf/analytics/statistics/WorksetCalculator.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * @brief Utility class to calculate workset from inserted ranges
+ *
+ * Workset is defined here as a length of union of all inputed ranges.
+ * Ranges are defined using beginning number and length similarily to how I/O
+ * requests are defined.
+ */
+class WorksetCalculator {
+public:
+    WorksetCalculator();
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.h]: probably we need copy constructor and assignment operator

File: source/octf/analytics/statistics/IoStatistics.cpp [hunk 13]
```diff
@@ -18,7 +18,8 @@ struct IoStatistics::Stats {
     Stats(const Stats &other)
             : SizeDistribution(other.SizeDistribution)
             , LatencyDistribution(other.LatencyDistribution)
-            , Errors(other.Errors) {}
+            , Errors(other.Errors)
+            , Wc() {}
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/IoStatistics.cpp]: other.Wc

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 14]
```diff
@@ -63,6 +63,14 @@ bool WorksetCalculator::Range::operator<(const Range &right) const {
 WorksetCalculator::WorksetCalculator()
         : m_hitRanges() {}
 
+WorksetCalculator::WorksetCalculator(const WorksetCalculator &)
+        : m_hitRanges() {}
+
+WorksetCalculator &octf::WorksetCalculator::operator=(
+        const WorksetCalculator &other) {
+    m_hitRanges = other.m_hitRanges;
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: Pls add compare between this and other

File: source/octf/analytics/statistics/WorksetCalculator.cpp [hunk 15]
```diff
@@ -63,6 +63,14 @@ bool WorksetCalculator::Range::operator<(const Range &right) const {
 WorksetCalculator::WorksetCalculator()
         : m_hitRanges() {}
 
+WorksetCalculator::WorksetCalculator(const WorksetCalculator &)
+        : m_hitRanges() {}
```
Comment (mariuszbarczak) [inline source/octf/analytics/statistics/WorksetCalculator.cpp]: m_hitRanges(other.m_hitRanges)

File: source/octf/analytics/statistics/IoStatistics.cpp (hunk 13 above)
Comment (mariuszbarczak) [reply source/octf/analytics/statistics/IoStatistics.cpp]: Also in assignment operator you need to consider Wc member.

**Pull request merged**